parquet-column/src/main/java/org/apache/parquet/io/MessageColumnIO.java

```diff
@@ -294,17 +301,33 @@ private void writeNullForMissingFieldsAtCurrentLevel() {
 
     private void writeNull(ColumnIO undefinedField, int r, int d) {
       if (undefinedField.getType().isPrimitive()) {
-        columnWriter[((PrimitiveColumnIO)undefinedField).getId()].writeNull(r, d);
+        columnWriter[((PrimitiveColumnIO) undefinedField).getId()].writeNull(r, d);
       } else {
-        GroupColumnIO groupColumnIO = (GroupColumnIO)undefinedField;
-        writeNullToLeaves(groupColumnIO, r, d);
+        GroupColumnIO groupColumnIO = (GroupColumnIO) undefinedField;
+        // only cache the repetition level, the definition level should always be the definition level of the parent node
+        cacheNullForGroup(groupColumnIO, r);
       }
     }
 
-    private void writeNullToLeaves(GroupColumnIO group, int r, int d) {
-      for(ColumnWriter leafWriter: groupToLeafWriter.get(group)) {
-        leafWriter.writeNull(r,d);
+    private void cacheNullForGroup(GroupColumnIO group, int r) {
+      if (!groupNullCache.containsKey(group)) {
```

Comment:
this is in the hot path, so we should make this as efficient as possible, so:
```
IntArrayList key = groupNullCache.get(group)
if (groupNullCache == null) {
  key = new IntArrayList()
  groupNullCache.put(group, key);
}
key.add(r);
```

This does only 1 map lookup when the key is already present, and 1 lookup, 1 put when it isn't (which will only happen once anyway)